src/awkward/_v2/record.py

```diff
@@ -191,11 +191,38 @@ def packed(self):
             return Record(self._array[self._at : self._at + 1].packed(), 0)
 
     def to_list(self, behavior=None):
-        cls = ak._v2._util.recordclass(self._array, behavior)
-        if cls is not ak._v2.highlevel.Record:
-            return cls(self)
+        return self._to_list(behavior, None)
 
-        return self._array[self._at : self._at + 1]._to_list(behavior, None)[0]
+    def _to_list(self, behavior, json_conversions):
```

Comment:
If we decide to use a flag, then I'd suggest also adding a decorator to set it, e.g.

```python
def ignore_in_to_list(func):
    func.ignore_in_to_list = True
    return func
```

I think the challenge here is working out what `to_list` is supposed to allow. I had originally been thinking that the slow case allows users to return non-awkward types, which would present a problem when using `from_iter` to re-load the same data (e.g. if it returns a random Python object). However, whilst it looks like we currently do permit that, I don't know whether we actively intend to support it. It might just be a side effect of allowing `to_list` to return something that is different to the default `to_list`, e.g. a filtered view of the data.


You seem to acknowledge this question in 
> this seems to me like a different thing, a performance hint instead of a behavior-changing thing.

So I'd be interested to hear your take. 

I don't know if I've already said this, but I'm wondering whether we need to do any of this - why not leave overriding `__getitem__` unsupported (given that strings are already handled internally)?  We could look to adding support down the road if someone needs it. Whereas, if we keep this code in-place, then we are accepting a burden of supporting it for the indefinite future.

So, take my opinion as "can we just remove this code?", but I'm on-board with adding a performance flag rather than full `to_list` overload if removal is off the table :)

Reply:
> So, take my opinion as "can we just remove this code?"

Well, the first thing is that we can't _prevent_ people from overriding `__getitem__`, and ignoring the fact that they might have overridden it would make the output of `to_list` wrong in a way that exposes its implementation. I can't see how we can _not_ do this.

If someone naively overrides `__getitem__` and `to_list` returns the right output but is slower, that's not an error. It might be undesirable ("might" depends on whether the user even cares about `to_list`, as that's a way of exiting Awkward's system and they might not plan to do that), but now we have a sneaky way to patch it, if the downstream developer can tell us that `__getitem__`'s behavior is not changed.

> I think the challenge here is working out what `to_list` is supposed to allow.

The `to_list` function converts Awkward data into Python data, as though iterating recursively, turning every Awkward list into a Python `list` and every Awkward record into a Python `dict` (and missing values into `None`). This is a naive implementation of `to_list`:

```python
>>> def to_list(array):
...     if isinstance(array, ak._v2.Array):
...         return [to_list(array[index]) for index in range(len(array))]
...     elif isinstance(array, ak._v2.Record):
...         return {field: to_list(array[field]) for field in array.fields}
...     else:
...         return array
```

Leaf-node values, the NumpyArrays, return what NumPy returns for its `tolist`. For example, if the NumPy types are dates, NumPy returns `datetime.datetime` objects:

```python
>>> array = ak._v2.Array(np.array([0, 1, 2], "datetime64[Y]"))
>>> array
<Array [1970, 1971, 1972] type='3 * datetime64[Y]'>
>>> array.tolist()
[datetime.date(1970, 1, 1), datetime.date(1971, 1, 1), datetime.date(1972, 1, 1)]
```

which are, in my opinion, random Python objects. The IP address case we talked about is another one in which someone would want to override `__getitem__`:

```python
>>> import ipaddress
>>> class IPv4Array(ak._v2.Array):
...     def __getitem__(self, where):
...         out = super().__getitem__(where)
...         return ipaddress.ip_address(np.asarray(out).tobytes())
... 
>>> ak._v2.behavior["IPv4"] = IPv4Array
>>> array = ak._v2.with_parameter(
...     ak._v2.without_parameters(
...         ak._v2.Array([b"\x7f\x00\x00\x01", b"\xff\xff\xff\xff"])
...     ),
...     "__array__",
...     "IPv4",
... )
>>> array
<IPv4Array [[127, 0, 0, 1], [255, ..., 255]] type='2 * [var * uint8, paramete...'>
>>> array[0]
IPv4Address('127.0.0.1')
>>> array[1]
IPv4Address('255.255.255.255')
>>> array.tolist()
[IPv4Address('127.0.0.1'), IPv4Address('255.255.255.255')]
```

Here, instead of going to Python date objects, or Python strings as in the case of `__array__ = "string"`, they're going into `IPv4Address` objects. That's what `to_list` is supposed to do: give you Python objects.

> would present a problem when using `from_iter` to re-load the same data

This is not round-trippable with `from_iter`, but it never was: `to_list` is a lossy operation that loses the distinction between fixed-length lists and variable-length lists, as well as the types of any list contents without values. This is also true of NumPy's `tolist` and `array` constructor. These functions provide ways of getting data between vectorized form and Python form, but these forms are not isomorphic to each other.